rust/module.rs

```diff
@@ -315,6 +307,102 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
+    name: String,
+    author: Option<String>,
+    description: Option<String>,
+    alias: Option<String>,
+    params: Option<Group>,
+}
+
+impl ModuleInfo {
+    fn parse(it: &mut token_stream::IntoIter) -> Self {
+        let mut info = ModuleInfo::default();
+        // The first allowed position is 1, to prevent setting something twice
+        let mut pos = 0u8;
+
+        loop {
+            let name = match it.next() {
+                Some(TokenTree::Ident(ident)) => ident.to_string(),
+                Some(_) => panic!("Expected Ident or end"),
+                None => break,
+            };
+
+            assert_eq!(expect_punct(it), ':');
+
+            if name == "type" {
+                allowed_pos(&mut pos, 1, &name);
+                let type_ = expect_ident(it);
+                info.type_ = type_;
+            } else if name == "params" {
+                allowed_pos(&mut pos, 7, &name);
+                let params = expect_group(it);
+                info.params = Some(params);
+            } else {
+
+                let value = expect_byte_string(it);
+
+                match name.as_str() {
+                    "name" => {
+                        allowed_pos(&mut pos, 2, &name);
+                        info.name = value;
+                    },
+                    "author" => {
+                        allowed_pos(&mut pos, 3, &name);
+                        info.author = Some(value);
+                    }
+                    "description" => {
+                        allowed_pos(&mut pos, 4, &name);
+                        info.description = Some(value);
+                    },
+                    "license" => {
+                        allowed_pos(&mut pos, 5, &name);
+                        info.license = value;
+                    },
+                    "alias" => {
+                        allowed_pos(&mut pos, 6, &name);
+                        info.alias = Some(value);
+                    },
+                    "alias_rtnl_link" => {
+                        allowed_pos(&mut pos, 6, &name);
+                        info.alias = Some(format!("rtnl-link-{}", value));
+                    },
+                    _ => panic!("field '{}' is not supported by module", name),
+                }
+            }
+            assert_eq!(expect_punct(it), ',');
+        }
+
+        expect_end(it);
+
+        info.check_required();
+
+        info
+    }
+
+    fn check_required(&self) {
+        if self.type_.is_empty() {
+            panic!("'type' not specified in module macro");
+        }
+        if self.license.is_empty() {
+            panic!("'license' not specified in module macro");
+        }
+        if self.name.is_empty() {
+            panic!("'name' not specified in module macro");
+        }
+    }
+}
+
+fn allowed_pos(pos: &mut u8, allowed_pos: u8, name: &str) {
+    if *pos > allowed_pos {
+        panic!("'{}' is not allowed at this position", name);
+    }
+    *pos = allowed_pos;
+}
```

Comment:
This one could be inside the `impl`. Actually, it could be local lambda so that we only need to write `allowed_pos(x)`.

Reply:
I can't get my head wrapped around making it a lambda at the moment. will put it inside the `impl`.